index.js

```diff
@@ -938,6 +933,152 @@ DB.prototype._visitTrie = function (key, path, node, heads, halt, visited, cb, t
   }
 }
 
+DB.prototype.createHistoryStream = function (opts) {
+  opts = opts || {}
+
+  var self = this
+
+  var stream = new Readable({objectMode: true})
+  stream._read = noop
+
+  function cb (err) {
+    stream.emit('error', err)
+  }
+
+  var nodes = []
+  var seqs = []
+  var atFront = false  // whether the live stream has reached real-time
+
+  // Populate 'nodes' and 'seq' with first entry in each writer
+  initTraversal(function (err, theNodes, theSeqs) {
+    if (err) return cb(err)
+    nodes = theNodes
+    seqs = theSeqs
+    traverseNode()
+  })
+
+  function traverseNode () {
+    var next = nodes.reduce(
+      function (a, b) {
+        if (!a && !b) return null
+        else if (!a) return b
+        else if (!b) return a
+        else if (a.seq <= b.clock[a.feed]) return a
+        else return b
+      }, null)
+
+    // Avoid ending the stream with a null value if opts.live
+    if (!opts.live || next !== null) stream.push(next)
+
+    // No new nodes to traverse
+    if (!next) {
+      // Stream has ended
+      if (atFront || !opts.live) return
+
+      // Stop traversal; set up event listeners; wait
+      atFront = true
+      self.once('_writer', onWriter)
+      self.once('append', onAppend)
+      return
+    }
+
+    if (next.seq === self._writers[next.feed].feed.length - 1) {
+      nodes[next.feed] = null
+      return traverseNode()
+    }
+
+    self._writers[next.feed].get(next.seq + 1, function (err, newNode) {
+      if (err) return cb(err)
+      nodes[next.feed] = newNode
+      seqs[next.feed] = newNode.seq
+      traverseNode()
+    })
+  }
+
+  function getStartSeq (heads, key) {
+    if (!heads) {
+      return 0
+    } else {
+      for (var i = 0; i < heads.length; i++) {
+        if (heads[i].key.equals(key)) {
+          return heads[i].seq + 1
+        }
+      }
+      return 0
+    }
+  }
+
+  function initTraversal (cb) {
+    var startHeads = opts.start ? versionToHeads(opts.start) : null
+    var nodes = []
+    var seqs = []
+
+    var pending = self._writers.length + 1
+    for (var i = 0; i < self._writers.length; i++) {
+      (function (n) {
```

Comment:
Why are you creating an anonymous function for every loop?
Sorry for the unsolicited comment @noffle - I have just been looking at your implementation in the light of reading your architecture description to see if it starts to make more sense. It does.

Reply:
Because `i` is lexicographically scoped. This is one of those places where `let` makes more sense :)

Reply:
Ah makes sense - thanks @mafintosh.
This is what arrow functions are for too as they retain the scope its parents context.